File: 2013/Round 1A/A. Bullseye/Bullseye.py
# Google Code Jam
# Round 1A 2013
# A. Bullseye
# https://code.google.com/codejam/contest/2418487/dashboard#s=p0
#
# Binary search solution in Python.
#
# Author: Sergey Dymchenko <kit1980@gmail.com>
#
# Python 2.6.5

def paint(r, n):
    return 2*r*n + 2*n*n - n

def find_max(r, t):
    lo = 0
    hi = t
    while lo < hi:
        mid = 1 + (lo + hi) // 2
        if paint(r, mid) <= t:
            lo = mid
        else:
            hi = mid - 1
    return lo

def do_case(case_num, r, t):
    result = find_max(r, t)
    print "Case #%d: %d" % (case_num, result)

def main():
    T = int(raw_input())
    for case_num in range(1, T + 1):
        r, t = map(int, raw_input().split())
        do_case(case_num, r, t)

if __name__ == "__main__":
    main()

File: 2013/Round 1A/A. Bullseye/Bullseye.pl
% Google Code Jam
% Round 1A 2013
% A. Bullseye
% https://code.google.com/codejam/contest/2418487/dashboard#s=p0
%
% Binary search solution in B-Prolog.
%
% Author: Sergey Dymchenko <kit1980@gmail.com>
%
% B-Prolog 8.0#1 - http://www.probp.com/
% Usage:
% sed 's/^ */[/; s/ *$/]./; s/ \+/, /g' in-file | bp -g "['Bullseye.pl'],main,halt" -l | tail -n +5 > out-file

paint(R, N, Paint) :-
    Paint is 2*R*N + 2*N*N - N.

find_max(_R, _T, N, N, N).
find_max(R, T, Lo, Hi, N) :-
    Lo < Hi,
    Mid is 1 + (Lo + Hi) // 2,
    paint(R, Mid, Paint),
    ( Paint =< T ->
        find_max(R, T, Mid, Hi, N)
    ;
        Mid1 is Mid - 1,
        find_max(R, T, Lo, Mid1, N)
    ).
    
do_case(Case_num, R, T) :-
    find_max(R, T, 0, T, N),
    format("Case #~w: ~w\n", [Case_num, N]).

main :-
    read([C]),
    foreach(Case_num in 1..C, [R, T], (
            read([R, T]),
            do_case(Case_num, R, T) 
        )).
